File: src/test/java/com/example/pages/SignupPage.java
package com.example.pages;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.support.ui.ExpectedConditions;

/**
 * Page Object for registration form – "Enter Account Information" section and address details.
 */
public class SignupPage extends BasePage {

    // ===== "Enter Account Information" section =====
    // <b>Enter Account Information</b>
    private final By enterAccountInformationSection =
            By.xpath("//*[contains(text(),'Enter Account Information')]");

    // Title: Mr / Mrs
    private final By titleMr = By.id("id_gender1");
    private final By titleMrs = By.id("id_gender2");

    // Form fields (data-qa)
    private final By passwordInput = By.cssSelector("[data-qa='password']");
    private final By daysSelect = By.cssSelector("[data-qa='days']");
    private final By monthsSelect = By.cssSelector("[data-qa='months']");
    private final By yearsSelect = By.cssSelector("[data-qa='years']");

    // Checkboxes
    private final By newsletterCheckbox = By.id("newsletter");
    private final By optinCheckbox = By.id("optin");

    // ===== Address fields (typical data-qa on automationexercise) =====
    private final By firstNameInput = By.cssSelector("[data-qa='first_name']");
    private final By lastNameInput = By.cssSelector("[data-qa='last_name']");
    private final By address1Input = By.cssSelector("[data-qa='address']");
    private final By countrySelect = By.cssSelector("[data-qa='country']");
    private final By stateInput = By.cssSelector("[data-qa='state']");
    private final By cityInput = By.cssSelector("[data-qa='city']");
    private final By zipcodeInput = By.cssSelector("[data-qa='zipcode']");
    private final By mobileNumberInput = By.cssSelector("[data-qa='mobile_number']");

    // Buttons
    private final By createAccountButton = By.cssSelector("[data-qa='create-account']");

    public SignupPage(WebDriver driver) {
        super(driver);
    }

    // ===== Section verification =====

    public boolean isEnterAccountInformationSectionVisible() {
        return isElementPresent(enterAccountInformationSection);
    }

    // ===== Filling account data =====

    public void fillAccountInformation(String password, String title, String day, String month, String year) {
        if ("Mrs".equalsIgnoreCase(title)) {
            click(titleMrs);
        } else {
            click(titleMr);
        }
        writeText(passwordInput, password);
        selectByValue(daysSelect, day);
        selectByValue(monthsSelect, month);
        selectByValue(yearsSelect, year);
    }

    public void subscribeToNewsletter() {
        if (!isCheckboxSelected(newsletterCheckbox)) {
            clickViaJavaScript(newsletterCheckbox);
        }
    }

    public void agreeToSpecialOffers() {
        if (!isCheckboxSelected(optinCheckbox)) {
            clickViaJavaScript(optinCheckbox);
        }
    }

    public void fillAddressDetails(
            String firstName,
            String lastName,
            String address1,
            String country,
            String state,
            String city,
            String zipcode,
            String mobileNumber
    ) {
        writeText(firstNameInput, firstName);
        writeText(lastNameInput, lastName);
        writeText(address1Input, address1);
        selectByVisibleText(countrySelect, country);
        writeText(stateInput, state);
        writeText(cityInput, city);
        writeText(zipcodeInput, zipcode);
        writeText(mobileNumberInput, mobileNumber);
    }

    public void clickCreateAccount() {
        wait.until(ExpectedConditions.elementToBeClickable(createAccountButton));
        clickViaJavaScript(createAccountButton);
    }
}

File: pages/signup_page.py
"""Signup page – Enter Account Information and address details."""
from selenium.webdriver.common.by import By
from selenium.webdriver.remote.webdriver import WebDriver
from selenium.webdriver.support import expected_conditions as EC

from pages.base_page import BasePage


class SignupPage(BasePage):
    enter_account_information_section = (By.XPATH, "//*[contains(text(),'Enter Account Information')]")
    title_mr = (By.ID, "id_gender1")
    title_mrs = (By.ID, "id_gender2")
    password_input = (By.CSS_SELECTOR, "[data-qa='password']")
    days_select = (By.CSS_SELECTOR, "[data-qa='days']")
    months_select = (By.CSS_SELECTOR, "[data-qa='months']")
    years_select = (By.CSS_SELECTOR, "[data-qa='years']")
    newsletter_checkbox = (By.ID, "newsletter")
    optin_checkbox = (By.ID, "optin")
    first_name_input = (By.CSS_SELECTOR, "[data-qa='first_name']")
    last_name_input = (By.CSS_SELECTOR, "[data-qa='last_name']")
    address1_input = (By.CSS_SELECTOR, "[data-qa='address']")
    country_select = (By.CSS_SELECTOR, "[data-qa='country']")
    state_input = (By.CSS_SELECTOR, "[data-qa='state']")
    city_input = (By.CSS_SELECTOR, "[data-qa='city']")
    zipcode_input = (By.CSS_SELECTOR, "[data-qa='zipcode']")
    mobile_number_input = (By.CSS_SELECTOR, "[data-qa='mobile_number']")
    create_account_button = (By.CSS_SELECTOR, "[data-qa='create-account']")

    def __init__(self, driver: WebDriver):
        super().__init__(driver)

    def is_enter_account_information_section_visible(self) -> bool:
        return self.is_element_present(self.enter_account_information_section)

    def fill_account_information(self, password: str, title: str, day: str, month: str, year: str) -> None:
        if title and title.lower() == "mrs":
            self.click(self.title_mrs)
        else:
            self.click(self.title_mr)
        self.write_text(self.password_input, password)
        self.select_by_value(self.days_select, day)
        self.select_by_value(self.months_select, month)
        self.select_by_value(self.years_select, year)

    def subscribe_to_newsletter(self) -> None:
        if not self.is_checkbox_selected(self.newsletter_checkbox):
            self.click_via_javascript(self.newsletter_checkbox)

    def agree_to_special_offers(self) -> None:
        if not self.is_checkbox_selected(self.optin_checkbox):
            self.click_via_javascript(self.optin_checkbox)

    def fill_address_details(
        self,
        first_name: str,
        last_name: str,
        address1: str,
        country: str,
        state: str,
        city: str,
        zipcode: str,
        mobile_number: str,
    ) -> None:
        self.write_text(self.first_name_input, first_name)
        self.write_text(self.last_name_input, last_name)
        self.write_text(self.address1_input, address1)
        self.select_by_visible_text(self.country_select, country)
        self.write_text(self.state_input, state)
        self.write_text(self.city_input, city)
        self.write_text(self.zipcode_input, zipcode)
        self.write_text(self.mobile_number_input, mobile_number)

    def click_create_account(self) -> None:
        self.wait.until(EC.element_to_be_clickable(self.create_account_button))
        self.click_via_javascript(self.create_account_button)
